File: src/desktop-edge-agent/src/FccDesktopAgent.Core/WebSocket/OdooWsModels.cs
using System.Text.Json.Serialization;
using FccDesktopAgent.Core.Adapter.Common;
using FccDesktopAgent.Core.Buffer.Entities;

namespace FccDesktopAgent.Core.WebSocket;

/// <summary>
/// Transaction DTO matching the legacy DOMSRealImplementation PumpTransactions model.
/// JSON field names MUST match the legacy contract exactly — Odoo POS parses them by name.
/// </summary>
public sealed class PumpTransactionWsDto
{
    [JsonPropertyName("id")]
    public int Id { get; set; }

    [JsonPropertyName("transaction_id")]
    public string TransactionId { get; set; } = string.Empty;

    [JsonPropertyName("pump_id")]
    public int PumpId { get; set; }

    [JsonPropertyName("nozzle_id")]
    public int NozzleId { get; set; }

    [JsonPropertyName("attendant")]
    public string? Attendant { get; set; }

    [JsonPropertyName("product_id")]
    public string ProductId { get; set; } = string.Empty;

    [JsonPropertyName("qty")]
    public decimal Qty { get; set; }

    [JsonPropertyName("unit_price")]
    public decimal UnitPrice { get; set; }

    [JsonPropertyName("total")]
    public decimal Total { get; set; }

    [JsonPropertyName("state")]
    public string State { get; set; } = "pending";

    [JsonPropertyName("start_time")]
    public string StartTime { get; set; } = string.Empty;

    [JsonPropertyName("end_time")]
    public string EndTime { get; set; } = string.Empty;

    [JsonPropertyName("order_uuid")]
    public string? OrderUuid { get; set; }

    [JsonPropertyName("sync_status")]
    public int SyncStatus { get; set; }

    [JsonPropertyName("odoo_order_id")]
    public string? OdooOrderId { get; set; }

    [JsonPropertyName("add_to_cart")]
    public bool AddToCart { get; set; }

    [JsonPropertyName("payment_id")]
    public string? PaymentId { get; set; }
}

/// <summary>
/// Fuel pump status DTO matching the legacy FuelPumpStatusDto.
/// WARNING: mixed casing is intentional — legacy contract uses snake_case, PascalCase, and camelCase.
/// </summary>
public sealed class FuelPumpStatusWsDto
{
    [JsonPropertyName("pump_number")]
    public int PumpNumber { get; set; }

    [JsonPropertyName("nozzle_number")]
    public int NozzleNumber { get; set; }

    [JsonPropertyName("status")]
    public string Status { get; set; } = "idle";

    [JsonPropertyName("reading")]
    public decimal Reading { get; set; }

    [JsonPropertyName("volume")]
    public decimal Volume { get; set; }

    [JsonPropertyName("litre")]
    public decimal Litre { get; set; }

    [JsonPropertyName("amount")]
    public decimal Amount { get; set; }

    [JsonPropertyName("attendant")]
    public string? Attendant { get; set; }

    [JsonPropertyName("count")]
    public int Count { get; set; }

    [JsonPropertyName("FpGradeOptionNo")]
    public int FpGradeOptionNo { get; set; }

    [JsonPropertyName("unit_price")]
    public decimal? UnitPriceValue { get; set; }

    [JsonPropertyName("isOnline")]
    public bool IsOnline { get; set; }
}

/// <summary>
/// WebSocket error response.
/// </summary>
public sealed class WsErrorResponse
{
    [JsonPropertyName("status")]
    public string Status { get; set; } = "error";

    [JsonPropertyName("message")]
    public string Message { get; set; } = string.Empty;
}

/// <summary>
/// Attendant pump count update item (inbound from Odoo).
/// </summary>
public sealed class AttendantPumpCountUpdateItem
{
    public int PumpNumber { get; set; }
    public string EmpTagNo { get; set; } = string.Empty;
    public int NewMaxTransaction { get; set; }
}

// ---------------------------------------------------------------------------
// Mapping extensions
// ---------------------------------------------------------------------------

public static class WsDtoMappers
{
    private static int _txIdCounter;

    /// <summary>
    /// Map a <see cref="BufferedTransaction"/> to the legacy <see cref="PumpTransactionWsDto"/>.
    /// </summary>
    public static PumpTransactionWsDto ToWsDto(this BufferedTransaction tx)
    {
        var qty = tx.VolumeMicrolitres / 1_000_000m;
        var price = tx.UnitPriceMinorPerLitre / 100m;
        var total = tx.AmountMinorUnits / 100m;

        var wsState = tx.IsDiscard ? "discard"
            : tx.SyncStatus is Adapter.Common.SyncStatus.SyncedToOdoo or Adapter.Common.SyncStatus.Uploaded
                ? "approved"
                : "pending";

        var wsSyncStatus = tx.SyncStatus == Adapter.Common.SyncStatus.Pending ? 0 : 1;

        return new PumpTransactionWsDto
        {
            Id = Interlocked.Increment(ref _txIdCounter),
            TransactionId = tx.FccTransactionId,
            PumpId = tx.PumpNumber,
            NozzleId = tx.NozzleNumber,
            Attendant = tx.AttendantId,
            ProductId = tx.ProductCode,
            Qty = qty,
            UnitPrice = price,
            Total = total,
            State = wsState,
            StartTime = tx.StartedAt.ToString("o"),
            EndTime = tx.CompletedAt.ToString("o"),
            OrderUuid = tx.OrderUuid,
            SyncStatus = wsSyncStatus,
            OdooOrderId = tx.OdooOrderId,
            AddToCart = tx.AddToCart,
            PaymentId = tx.PaymentId,
        };
    }

    /// <summary>
    /// Map a <see cref="PumpStatus"/> to the legacy <see cref="FuelPumpStatusWsDto"/>.
    /// </summary>
    public static FuelPumpStatusWsDto ToWsDto(this PumpStatus ps)
    {
        _ = decimal.TryParse(ps.CurrentVolumeLitres, out var volume);
        _ = decimal.TryParse(ps.CurrentAmount, out var amount);
        decimal? price = decimal.TryParse(ps.UnitPrice, out var p) ? p : null;

        var status = ps.State switch
        {
            PumpState.Idle => "idle",
            PumpState.Authorized => "authorized",
            PumpState.Calling => "calling",
            PumpState.Dispensing => "dispensing",
            PumpState.Paused => "suspended",
            PumpState.Completed => "idle",
            PumpState.Error => "inoperative",
            PumpState.Offline => "offline",
            _ => "unknown",
        };

        return new FuelPumpStatusWsDto
        {
            PumpNumber = ps.PumpNumber,
            NozzleNumber = ps.NozzleNumber,
            Status = status,
            Reading = 0,
            Volume = volume,
            Litre = volume,
            Amount = amount,
            Attendant = null,
            Count = 0,
            FpGradeOptionNo = 0,
            UnitPriceValue = price,
            IsOnline = ps.State is not PumpState.Offline and not PumpState.Error,
        };
    }
}

File: src/edge-agent/app/src/main/kotlin/com/fccmiddleware/edge/websocket/OdooWsModels.kt
package com.fccmiddleware.edge.websocket

import com.fccmiddleware.edge.buffer.dao.WsBufferedTransaction
import com.fccmiddleware.edge.buffer.entity.BufferedTransaction
import com.fccmiddleware.edge.adapter.common.CurrencyUtils
import com.fccmiddleware.edge.adapter.common.PumpStatus
import com.fccmiddleware.edge.adapter.common.PumpState
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

// ---------------------------------------------------------------------------
// Outbound DTOs — sent TO Odoo POS over WebSocket
// ---------------------------------------------------------------------------

/**
 * Transaction DTO matching the legacy DOMSRealImplementation PumpTransactions model.
 *
 * JSON field names MUST match the legacy contract exactly — Odoo POS parses these
 * by name and any deviation breaks backward compatibility.
 */
@Serializable
data class PumpTransactionWsDto(
    @SerialName("id") val id: Int,
    @SerialName("transaction_id") val transactionId: String,
    @SerialName("pump_id") val pumpId: Int,
    @SerialName("nozzle_id") val nozzleId: Int,
    @SerialName("attendant") val attendant: String?,
    @SerialName("product_id") val productId: String,
    @SerialName("qty") val qty: Double,
    @SerialName("unit_price") val unitPrice: Double,
    @SerialName("total") val total: Double,
    @SerialName("state") val state: String,
    @SerialName("start_time") val startTime: String,
    @SerialName("end_time") val endTime: String,
    @SerialName("order_uuid") val orderUuid: String?,
    @SerialName("sync_status") val syncStatus: Int,
    @SerialName("odoo_order_id") val odooOrderId: String?,
    @SerialName("add_to_cart") val addToCart: Boolean,
    @SerialName("payment_id") val paymentId: String?,
)

/**
 * Fuel pump status DTO matching the legacy FuelPumpStatusDto.
 *
 * WARNING: mixed casing is intentional — legacy contract uses:
 * - snake_case: pump_number, nozzle_number, unit_price
 * - PascalCase: FpGradeOptionNo
 * - camelCase: isOnline
 * Odoo POS reads these field names directly. Do NOT normalise casing.
 */
@Serializable
data class FuelPumpStatusWsDto(
    @SerialName("pump_number") val pumpNumber: Int,
    @SerialName("nozzle_number") val nozzleNumber: Int,
    @SerialName("status") val status: String,
    @SerialName("reading") val reading: Double,
    @SerialName("volume") val volume: Double,
    @SerialName("litre") val litre: Double,
    @SerialName("amount") val amount: Double,
    @SerialName("attendant") val attendant: String?,
    @SerialName("count") val count: Int,
    @SerialName("FpGradeOptionNo") val fpGradeOptionNo: Int,
    @SerialName("unit_price") val unitPrice: Double?,
    @SerialName("isOnline") val isOnline: Boolean,
)

/**
 * WebSocket error response.
 */
@Serializable
data class WsErrorResponse(
    val status: String = "error",
    val message: String,
)

/**
 * Attendant pump count update acknowledgment.
 */
@Serializable
data class WsAttendantPumpCountAck(
    @SerialName("pump_number") val pumpNumber: Int,
    @SerialName("emp_tag_no") val empTagNo: String,
    @SerialName("max_limit") val maxLimit: Int,
    @SerialName("status") val status: String,
)

// ---------------------------------------------------------------------------
// Inbound message — parsed from the `mode` field
// ---------------------------------------------------------------------------

/**
 * Inbound attendant pump count update item.
 */
@Serializable
data class AttendantPumpCountUpdateItem(
    @SerialName("PumpNumber") val pumpNumber: Int,
    @SerialName("EmpTagNo") val empTagNo: String,
    @SerialName("NewMaxTransaction") val newMaxTransaction: Int,
)

// ---------------------------------------------------------------------------
// Mapping helpers
// ---------------------------------------------------------------------------

/**
 * Auto-incrementing counter for legacy integer `id` field.
 * AF-048: Initialized from epoch seconds so IDs never collide with a previous session
 * after process restart. Monotonically increments within a session.
 */
private val txIdCounter = java.util.concurrent.atomic.AtomicInteger(
    (System.currentTimeMillis() / 1000).toInt()
)

/**
 * Map a [BufferedTransaction] to the legacy [PumpTransactionWsDto] format.
 *
 * Monetary conversion: minor units → major units (divide by currency factor).
 * Volume conversion: microlitres → litres (divide by 1,000,000).
 */
fun BufferedTransaction.toWsDto(): PumpTransactionWsDto {
    val qty = volumeMicrolitres / 1_000_000.0
    val currencyFactor = CurrencyUtils.getFactor(currencyCode)
    val price = unitPriceMinorPerLitre / currencyFactor
    val total = amountMinorUnits / currencyFactor

    val wsState = when {
        isDiscard -> "discard"
        syncStatus == "SYNCED_TO_ODOO" -> "approved"
        syncStatus == "UPLOADED" -> "approved"
        else -> "pending"
    }
    val wsSyncStatus = if (syncStatus == "PENDING") 0 else 1

    return PumpTransactionWsDto(
        id = txIdCounter.incrementAndGet(),
        transactionId = fccTransactionId,
        pumpId = pumpNumber,
        nozzleId = nozzleNumber,
        attendant = attendantId,
        productId = productCode,
        qty = qty,
        unitPrice = price,
        total = total,
        state = wsState,
        startTime = startedAt,
        endTime = completedAt,
        orderUuid = orderUuid,
        syncStatus = wsSyncStatus,
        odooOrderId = odooOrderId,
        addToCart = addToCart,
        paymentId = paymentId,
    )
}

/**
 * AP-025: Map a [WsBufferedTransaction] projection to the legacy [PumpTransactionWsDto] format.
 * Same logic as [BufferedTransaction.toWsDto] but operates on the lightweight projection
 * that excludes rawPayloadJson.
 */
fun WsBufferedTransaction.toWsDto(): PumpTransactionWsDto {
    val qty = volumeMicrolitres / 1_000_000.0
    val currencyFactor = CurrencyUtils.getFactor(currencyCode)
    val price = unitPriceMinorPerLitre / currencyFactor
    val total = amountMinorUnits / currencyFactor

    val wsState = when {
        isDiscard -> "discard"
        syncStatus == "SYNCED_TO_ODOO" -> "approved"
        syncStatus == "UPLOADED" -> "approved"
        else -> "pending"
    }
    val wsSyncStatus = if (syncStatus == "PENDING") 0 else 1

    return PumpTransactionWsDto(
        id = txIdCounter.incrementAndGet(),
        transactionId = fccTransactionId,
        pumpId = pumpNumber,
        nozzleId = nozzleNumber,
        attendant = attendantId,
        productId = productCode,
        qty = qty,
        unitPrice = price,
        total = total,
        state = wsState,
        startTime = startedAt,
        endTime = completedAt,
        orderUuid = orderUuid,
        syncStatus = wsSyncStatus,
        odooOrderId = odooOrderId,
        addToCart = addToCart,
        paymentId = paymentId,
    )
}

/**
 * Map a [PumpStatus] to the legacy [FuelPumpStatusWsDto] format.
 */
fun PumpStatus.toWsDto(): FuelPumpStatusWsDto {
    val volume = currentVolumeLitres?.toDoubleOrNull() ?: 0.0
    val amount = currentAmount?.toDoubleOrNull() ?: 0.0
    val price = unitPrice?.toDoubleOrNull()

    val statusStr = when (state) {
        PumpState.IDLE -> "idle"
        PumpState.AUTHORIZED -> "authorized"
        PumpState.CALLING -> "calling"
        PumpState.DISPENSING -> "dispensing"
        PumpState.PAUSED -> "suspended"
        PumpState.COMPLETED -> "idle"
        PumpState.ERROR -> "inoperative"
        PumpState.OFFLINE -> "offline"
        PumpState.UNKNOWN -> "unknown"
    }

    return FuelPumpStatusWsDto(
        pumpNumber = pumpNumber,
        nozzleNumber = nozzleNumber,
        status = statusStr,
        reading = 0.0,
        volume = volume,
        litre = volume,
        amount = amount,
        attendant = null,
        count = 0,
        fpGradeOptionNo = 0,
        unitPrice = price,
        isOnline = state != PumpState.OFFLINE && state != PumpState.ERROR,
    )
}
